# Co-expression Network Analysis of Susceptibility to Wheat Stem Rust

by Rafael Della Coletta and Cory Hirsch (September 2020).

---

The objective of this project is to create gene co-expression networks from wheat and Brachypodium genes in response to wheat stem rust infection.



## Requirements

| Software | Version | Additional libraries / modules |
| -------- | ------- | ------------------------------ |
| R        |         |                                |
| Python   |         |                                |
| Camoco   |         |                                |



## Data

All data, scripts, and output of analyses are located on the folder `/home/hirschc1/della028/projects/stem-rust_networks` from my account at the Minnesota Supercomputing Institute (MSI).

```bash
cd ~/projects/

mkdir -p stem-rust_networks/{analysis,data,scripts}
```

The lead author of the paper, Eva Henningsen, shared with me all data necessary for building the networks and they were stored at `data` folder.

* Wheat annotation file (`IWGSC_v1.1_HC_20170706.gff3`)
* Brachypodium annotation file (`BdistachyonBd21_3_537_v1.2.gene_exons.gff3`)
* Normalized and raw gene expression matrices for wheat and Brachy (`EH_brachy_counts_normalized.txt`, `EH_wheat_counts_normalized.txt`, `wheat_counts_raw.txt`, `brachy_counts_raw.txt`)
* GO terms for wheat and Brachy (`wheat_full_gos_long.txt`, `brachy_full_gos_long.txt`)



### Normalize raw expression by FPKM

Camoco requires that expression data is normalized by FPKM, but the normalized files Eva sent to me was generated by DEseq2 package from R, which is not FPKM. Thus, I wrote the script `scripts/counts2fpkm.R` to normalize the raw counts from HTseq into FPKM according to guidelines described by the [National Cancer Institute Genomic Data Commons](https://docs.gdc.cancer.gov/Data/Bioinformatics_Pipelines/Expression_mRNA_Pipeline/#mrna-expression-ht-seq-normalization). It uses the following formula to calculate FPKM:

```
                     (reads mapped to the gene) * 10^9
 FPKM = ------------------------------------------------------------
         (reads mapped to all protein-coding genes) * (gene length)
```

I ran `scripts/counts2fpkm.R` for both wheat and brachy datasets.

```bash
# first need to remove header from annotation
grep -v "^#" data/IWGSC_v1.1_HC_20170706.gff3 > data/IWGSC_v1.1_HC_20170706.no-header.gff3
grep -v "^#" data/BdistachyonBd21_3_537_v1.2.gene_exons.gff3 > data/BdistachyonBd21_3_537_v1.2.gene_exons.no-header.gff3

# then remove quotes around gene and sample names
sed -i 's/"//g' data/wheat_counts_raw.txt
sed -i 's/"//g' data/brachy_counts_raw.txt
# remove htseq-specific lines (i.e. start with _)
sed -i '/^_/d' data/wheat_counts_raw.txt
sed -i '/^_/d' data/brachy_counts_raw.txt

# for how to use script
Rscript scripts/counts2fpkm.R --help

# wheat dataset
Rscript scripts/counts2fpkm.R data/wheat_counts_raw.txt data/IWGSC_v1.1_HC_20170706.no-header.gff3 data/wheat_counts_fpkm.txt --cores=10

# brachy dataset
Rscript scripts/counts2fpkm.R data/brachy_counts_raw.txt data/BdistachyonBd21_3_537_v1.2.gene_exons.no-header.gff3 data/brachy_counts_fpkm.txt --cores=10
```

As a quick QC, I wrote `scripts/pca_expr_data.R` to perform PCA on the expression datasets (raw counts and fpkm) for wheat and Brachy.

```bash
mkdir -p analysis/qc

for geno in wheat brachy; do
  for expr in raw fpkm; do
    Rscript scripts/pca_expr_data.R data/${geno}_counts_${expr}.txt analysis/qc/pca_${geno}_${expr}.png
  done
done
```

> Grouping of samples according to PCA seems to agree reasonably well.



### Filter data by CV and samples for each network

Coexpression networks require variation of gene expression across samples. Thus, filtering genes by their coefficient of variation (CV) eliminates such uninformative genes, reduces the initial number of genes to build networks and speeds up clustering. To have an idea of what CV threshold to use, I plotted the distribution of genes CV among samples using `scripts/prepare_data_for_camoco.R`.

```bash
# plot CV distribution
Rscript scripts/prepare_data_for_camoco.R data/brachy_counts_fpkm.txt analysis/qc/cv_distribution_brachy-fpkm.png --plot-cv
Rscript scripts/prepare_data_for_camoco.R data/wheat_counts_fpkm.txt analysis/qc/cv_distribution_wheat-fpkm.png --plot-cv
```

Below is the number of genes that would be eliminated by different CV thresholds. Based on this table, I will only keep genes with CV >= 0.1 for both brachy and wheat.

|               | Brachy               | Wheat                |
| ------------- | -------------------- | -------------------- |
| Not expressed | 8398 genes (21.5%)   | 28795 genes (26.69%) |
| CV < 0.1      | 8518 genes (21.8%)   | 28797 genes (26.69%) |
| CV < 0.2      | 12678 genes (32.45%) | 30307 genes (28.09%) |
| CV < 0.3      | 18525 genes (47.42%) | 38032 genes (35.25%) |
| CV < 0.4      | 22511 genes (57.62%) | 47331 genes (43.87%) |
| CV < 0.5      | 25351 genes (64.89%) | 55537 genes (51.48%) |

An additional formatting of the expression matrix is required for building networks with Camoco. `scripts/prepare_data_for_camoco.R` generates a `.csv` file without the first comma from the header, in order to the input dataset look like this:

| Sample1 | Sample2 | Sample3 | Sample4 |    |
|---------|---------|---------|---------|----|
| gene1   | 10      | 20      | 30      | 40 |
| gene2   | 10      | 20      | 30      | 40 |
| gene3   | 10      | 20      | 30      | 40 |

I also to filter the whole dataset to have the samples I want for each network. There will be one network for each genotype (wheat susceptible, wheat resistant, brachy) using all reps for 2, 4 and 6 days after inoculating the plants with stem rust, and all reps for the 2dpi samples that were mock inoculated.

```bash
# brachy dataset
Rscript scripts/prepare_data_for_camoco.R data/brachy_counts_fpkm.txt data/expr_data_fpkm_brachy.cv_0.1.inf_2-4-6.mock_2.csv --filter-cv=0.1 --keep-samples=Bd21_D2_mock_R1,Bd21_D2_mock_R2,Bd21_D2_mock_R3,Bd21_D2_treated_R1,Bd21_D2_treated_R2,Bd21_D2_treated_R3,Bd21_D4_treated_R1,Bd21_D4_treated_R2,Bd21_D4_treated_R3,Bd21_D6_treated_R1,Bd21_D6_treated_R2,Bd21_D6_treated_R3

# wheat resistant dataset
Rscript scripts/prepare_data_for_camoco.R data/wheat_counts_fpkm.txt data/expr_data_fpkm_wheat_R.cv_0.1.inf_2-4-6.mock_2.csv --filter-cv=0.1 --keep-samples=Sr9b_D2_mock_R1,Sr9b_D2_mock_R2,Sr9b_D2_mock_R3,Sr9b_D2_treated_R1,Sr9b_D2_treated_R2,Sr9b_D2_treated_R3,Sr9b_D4_treated_R1,Sr9b_D4_treated_R2,Sr9b_D4_treated_R3,Sr9b_D6_treated_R1,Sr9b_D6_treated_R2,Sr9b_D6_treated_R3

# wheat susceptible dataset
Rscript scripts/prepare_data_for_camoco.R data/wheat_counts_fpkm.txt data/expr_data_fpkm_wheat_S.cv_0.1.inf_2-4-6.mock_2.csv --filter-cv=0.1 --keep-samples=W2691_D2_mock_R1,W2691_D2_mock_R2,W2691_D2_mock_R3,W2691_D2_treated_R1,W2691_D2_treated_R2,W2691_D2_treated_R3,W2691_D4_treated_R1,W2691_D4_treated_R2,W2691_D4_treated_R3,W2691_D6_treated_R1,W2691_D6_treated_R2,W2691_D6_treated_R3
```



## Installing Camoco

Camoco was already installed in a **virtual environment** using the software [miniconda](https://conda.io/miniconda.html).

```bash
cd ~/software/

# get miniconda bash installer
wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh
# install miniconda
bash Miniconda3-latest-Linux-x86_64.sh -p ~/software/miniconda
# check if conda needs uptdates
conda update -n base conda
# create virtual environment for camoco
conda create -n camoco python=3.6
# check installed environments
conda info --envs

# update pip, wheel and setuptools
pip install --upgrade pip
pip install --upgrade wheel
pip install --upgrade setuptools

# activate environment and install Camoco on that environment
source activate camoco
# notice that (camoco) now appears before "della028@ln0005 [~/software] %"
pip install numpy
pip install camoco
# test if package was installed correctly
camoco -h

# deactivate environment when done working
source deactivate
```



## Building networks

### Load reference annotation and gene ontology

Before building any network, I need to load Brachy and wheat reference genome annotations into Camoco's database. Once a reference genome is loaded, you don't need to load it again: it will stay in Camoco's database to be used as many time as you want.


```bash
# activate camoco virtual environment
source activate camoco

# create a reference genome dataset for camoco -- see 'camoco build-refgen -h' for help
camoco build-refgen data/BdistachyonBd21_3_537_v1.2.gene_exons.gff3 BrachyRef Bd21_3_537_v1.2 phytozomev13 Brachypodium_distachyon
camoco build-refgen data/IWGSC_v1.1_HC_20170706.gff3 WheatRef TraesChineseSpring_v1.1_201706 IWGSC Triticum_aestivum
# check that loading was successfull
camoco ls

# deactivate virtual environment
source deactivate
```

Camoco performs Gene Ontology (GO) enrichment analysis to quality control the networks created. In order to do that, I need to create a GO object in Camoco, which requires a `go.obo` file containing all core ontology terms (<http://geneontology.org/docs/download-ontology/>) and a two-column species-specific file relating genes (first column) and their respective GO terms (second column).

```bash
# activate camoco virtual environment
source activate camoco

# get go.obo file
wget -P data/ http://purl.obolibrary.org/obo/go.obo

# remove header from species go files
sed -i 1d data/brachy_full_gos_long.txt
sed -i 1d data/wheat_full_gos_long.txt

# need to add '.v1.2' at the end of brachy annotation gene IDs to match with reference gene IDs
sed -i 's/\t/.v1.2\t/g' data/brachy_full_gos_long.txt

# create a go annotation reference for camoco  -- see 'camoco build-go -h' for help
camoco build-go data/brachy_full_gos_long.txt data/go.obo BrachyGO brachy_go_annotation BrachyRef
camoco build-go data/wheat_full_gos_long.txt data/go.obo WheatGO wheat_go_annotation WheatRef
# check that loading was successfull
camoco ls

# deactivate virtual environment
source deactivate
```

> Done: Ontology:BrachyGO - desc: brachy_go_annotation - contains 3589 terms for Reference Genome: Brachypodium_distachyon - phytozomev13 - BrachyRef
> Done: Ontology:WheatGO - desc: wheat_go_annotation - contains 12572 terms for Reference Genome: Triticum_aestivum - IWGSC - WheatRef
