# Co-expression Network Analysis of Susceptibility to Wheat Stem Rust

by Rafael Della Coletta and Cory Hirsch (September 2020).

---

The objective of this project is to create gene co-expression networks from wheat and Brachypodium genes in response to wheat stem rust infection.



## Requirements

| Software | Version | Additional libraries / modules |
| -------- | ------- | ------------------------------ |
| R        |         |                                |
| Python   |         |                                |
| Camoco   |         |                                |



## Data

All data, scripts, and output of analyses are located on the folder `/home/hirschc1/della028/projects/stem-rust_networks` from my account at the Minnesota Supercomputing Institute (MSI).

```bash
cd ~/projects/

mkdir -p stem-rust_networks/{analysis,data,scripts}
```

The lead author of the paper, Eva Henningsen, shared with me all data necessary for building the networks and they were stored at `data` folder.

* Wheat annotation file (`IWGSC_v1.1_HC_20170706.gff3`)
* Brachypodium annotation file (`BdistachyonBd21_3_537_v1.2.gene_exons.gff3`)
* Normalized and raw gene expression matrices for wheat and Brachy (`EH_brachy_counts_normalized.txt`, `EH_wheat_counts_normalized.txt`, `wheat_counts_raw.txt`, `brachy_counts_raw.txt`)
* GO terms for wheat and Brachy (`wheat_full_gos_long.txt`, `brachy_full_gos_long.txt`)



### Normalize raw expression by FPKM

Camoco requires that expression data is normalized by FPKM, but the normalized files Eva sent to me was generated by DEseq2 package from R, which is not FPKM. Thus, I wrote the script `scripts/counts2fpkm.R` to normalize the raw counts from HTseq into FPKM according to guidelines described by the [National Cancer Institute Genomic Data Commons](https://docs.gdc.cancer.gov/Data/Bioinformatics_Pipelines/Expression_mRNA_Pipeline/#mrna-expression-ht-seq-normalization). It uses the following formula to calculate FPKM:

```
                     (reads mapped to the gene) * 10^9
 FPKM = ------------------------------------------------------------
         (reads mapped to all protein-coding genes) * (gene length)
```

I ran `scripts/counts2fpkm.R` for both wheat and brachy datasets.

```bash
# first need to remove header from annotation
grep -v "^#" data/IWGSC_v1.1_HC_20170706.gff3 > data/IWGSC_v1.1_HC_20170706.no-header.gff3
grep -v "^#" data/BdistachyonBd21_3_537_v1.2.gene_exons.gff3 > data/BdistachyonBd21_3_537_v1.2.gene_exons.no-header.gff3

# then remove quotes around gene and sample names
sed -i 's/"//g' data/wheat_counts_raw.txt
sed -i 's/"//g' data/brachy_counts_raw.txt
# remove htseq-specific lines (i.e. start with _)
sed -i '/^_/d' data/wheat_counts_raw.txt
sed -i '/^_/d' data/brachy_counts_raw.txt

# for how to use script
Rscript scripts/counts2fpkm.R --help

# wheat dataset
Rscript scripts/counts2fpkm.R data/wheat_counts_raw.txt data/IWGSC_v1.1_HC_20170706.no-header.gff3 data/wheat_counts_fpkm.txt --cores=10

# brachy dataset
Rscript scripts/counts2fpkm.R data/brachy_counts_raw.txt data/BdistachyonBd21_3_537_v1.2.gene_exons.no-header.gff3 data/brachy_counts_fpkm.txt --cores=10
```

As a quick QC, I wrote `scripts/pca_expr_data.R` to perform PCA on the expression datasets (raw counts and fpkm) for wheat and Brachy.

```bash
mkdir -p analysis/qc

for geno in wheat brachy; do
  for expr in raw fpkm; do
    Rscript scripts/pca_expr_data.R data/${geno}_counts_${expr}.txt analysis/qc/pca_${geno}_${expr}.png
  done
done
```

> Grouping of samples according to PCA seems to agree pretty well.
